### DISCUSSION TITLE
Creating multiple workers to do DQN

### DISCUSSION THREAD

**Discussion by raymondchua** [jax-ml/jax#17755] [Q&A]
Hi all, I hope this is the right place to ask this question. I would like be able to collect experience from 64 parallel environment copies into a single replay buffer. For exampling,  using 8 parallel workers, using shared model parameters and averaging gradients between workers. Does anyone know if this can be done in Jax using vmap and jit? Perhaps similar to this: https://github.com/google/jax/discussions/11565#discussioncomment-3194948

**Comment by anh-tong** [comment]
Hi, 

It seems like you can [`pmap`](https://jax.readthedocs.io/en/latest/jax-101/06-parallelism.html) in this case. In my own experience, I use some of Flax utility functions (I think you can write your own similar function with no difficulty).

For examples,
- [flax.jax_utils.replicate](https://flax.readthedocs.io/en/latest/api_reference/flax.jax_utils.html#flax.jax_utils.replicate) to make shared model parameters available to all devices
- [flax.training.common_utils.shard](https://github.com/google/flax/blob/63420d35d6d54f0158818f742ba206635a336e5b/flax/training/common_utils.py#L27) to share data to all devices
- `jax.lax.pmean`: use to average computed gradients


Here is a simple structure I use
```python
def compute_fn(params, input):
    grad = <compute grad here>
    
    # note that `axis_name` should be the same with `pmap`
    grad = jax.lax.pmean(grad, axis_name="batch")
    
    return grad

# make model available to all devices
params = <model>
params = replicate(params)

# make `pmap` version of `compute_fn`
p_compute_fn = jax.pmap(compute_fn, axis_name="batch", donate_argnums=(0,))

# shard data across all devices
data = <input data>
data = shard(data)

grad = p_compute_fn(params, data)

# `grad` is still replicated across the devices, can be retrieved as
grad = unreplicate(grad)
```

This is what I usually do. Maybe there is a better approach out there. 

Hope this could help.
